[ResourceMonitor] Add more meaningful release log information part 2.

basuke opened the pull request:
#### c7af5b9e0c4f785316499f4020ecd5dd8b4d4090
<pre>
[ResourceMonitor] Add more meaningful release log information part 2.
<a href="https://bugs.webkit.org/show_bug.cgi?id=288174">https://bugs.webkit.org/show_bug.cgi?id=288174</a>
<a href="https://rdar.apple.com/145259890">rdar://145259890</a>

Reviewed by Chris Dumez.

Adding new log category and move current release logs to that category. And also add more meaningful information
for the record.

This is part 2, targeting WebKit.

* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::networkUsageLevel const):
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::ResourceMonitorChecker::checkEligibility):
(WebCore::ResourceMonitorChecker::setContentRuleList):
(WebCore::ResourceMonitorChecker::finishPendingQueries):
(WebCore::ResourceMonitorChecker::setNetworkUsageThreshold):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::shouldOffloadIFrameForHost):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::resourceMonitorThrottler):
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::ResourceMonitorURLsController::prepare):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::loadOrUpdateResourceMonitorRuleList):
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::updateBytesTransferredOverNetwork):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setResourceMonitorContentRuleList):
(WebKit::WebProcess::setResourceMonitorContentRuleListAsync):

Canonical link: <a href="https://commits.webkit.org/292052@main">https://commits.webkit.org/292052@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb014aaf66c06ec2f0020f6a7e37923a7af3b25c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14140 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3985 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99568 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96598 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14438 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22568 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72148 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29464 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10752 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85404 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52480 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10447 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3105 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44387 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80663 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3211 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101611 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21604 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15761 "Found 1 new test failure: html5lib/generated/run-entities01-data.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81151 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81426 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80525 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20160 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25079 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2499 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14823 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21581 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26711 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21260 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24723 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22998 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -38,6 +38,13 @@ class LocalFrame;
+        Empty = 0,
+        Low = 20,
+        Medium = 60,
+        High = 80,
```
I am worried this is too much logging and I'm unclear how it will help us in sysdiagnoses. Why do we need 5 levels when this is basically a binary thing (too much data or not?)

I think we basically want logging when a frame gets unloaded due to large usage, that's it.

basuke replied:
I want to understand the actual usage in the real world web site. I need these value to understand how usage grows without logging whole usage. It creates max of 5 more logs for each frame so I don't think this causes too much logging. And also after understanding the usage characteristics at the end, we can hide this under compiler flag. I want to ask you to let this in the code.

cdumez commented on Source/WebCore/loader/ResourceMonitor.cpp:
```diff
@@ -154,6 +154,30 @@ void ResourceMonitor::addNetworkUsage(size_t bytes)
         checkNetworkUsageExcessIfNecessary();
 }
 
+size_t ResourceMonitor::networkUsageThreshold() const
```
Should be inlined in the header.

cdumez commented on Source/WebCore/loader/ResourceMonitor.h:
```diff
@@ -48,6 +55,8 @@ class ResourceMonitor final : public RefCountedAndCanMakeWeakPtr<ResourceMonitor
 
     void setDocumentURL(URL&&);
     WEBCORE_EXPORT void addNetworkUsage(size_t);
+    WEBCORE_EXPORT size_t networkUsageThreshold() const;
```
Should inline and drop `WEBCORE_EXPORT`

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -103,7 +101,7 @@ void ResourceMonitorChecker::setContentRuleList(ContentExtensions::ContentExtens
         m_ruleListIsPreparing = false;
 
-        RESOURCEMONITOR_RELEASE_LOG("Content rule list is set");
+        RELEASE_LOG(ResourceMonitoring, "ResourceMonitorChecker::Content rule list is set,");
```
What's the new `,` about?

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -90,7 +88,7 @@ ResourceMonitorChecker::Eligibility ResourceMonitorChecker::checkEligibility(con
 
     auto matched = m_ruleList->processContentRuleListsForResourceMonitoring(info.resourceURL, info.mainDocumentURL, info.frameURL, info.type);
-    RESOURCEMONITOR_RELEASE_LOG("The url is %" PUBLIC_LOG_STRING ": %" SENSITIVE_LOG_STRING " (%" PUBLIC_LOG_STRING ")", (matched ? "eligible" : "not eligible"), info.resourceURL.string().utf8().data(), ContentExtensions::resourceTypeToString(info.type).characters());
+    RELEASE_LOG(ResourceMonitoring, "ResourceMonitorChecker::The url is %" PUBLIC_LOG_STRING ": %" SENSITIVE_LOG_STRING " (%" PUBLIC_LOG_STRING ")", (matched ? "eligible" : "not eligible"), info.resourceURL.string().utf8().data(), ContentExtensions::resourceTypeToString(info.type).characters());
```
The prefix we usually use `ClassName::FunctionName: `. `ResourceMonitorChecker::The url` looks really weird. Same comment applies elsewhere in this PR.

cdumez commented on Source/WebCore/loader/ResourceMonitorChecker.cpp:
```diff
@@ -34,8 +34,6 @@
 
 #if ENABLE(CONTENT_EXTENSIONS)
 
-#define RESOURCEMONITOR_RELEASE_LOG(fmt, ...) RELEASE_LOG(ResourceMonitoring, "ResourceMonitorChecker::" fmt, ##__VA_ARGS__)
```
Not sure why we're dropping this convenience macro. It would avoid repeating `ResourceMonitorChecker::` every time.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -2037,6 +2037,10 @@ void WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold()
     if (url.isEmpty())
         return;
 
+    auto host = url.host().toStringWithoutCopying();
```
I don't think the local variable should be a String constructed from `.toStringWithoutCopying()`. This is too risky / error prone. When something comes from `toStringWithoutCopying()`, it should be used right away, on the same line, so we understand it is safe not to copy.

cdumez commented on LayoutTests/http/tests/security/mixedContent/resources/frame-with-insecure-localhost-image.html:
```diff
@@ -1,4 +1,5 @@
 <img src="http://localhost:8080/security/resources/compass.jpg">
+<img src="http://127.0.0.1:8080/security/resources/compass.jpg">
```
Why this change?

basuke replied:
Oh my. That's wrong one.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -2038,6 +2038,10 @@ void WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold()
 
+    auto host = url.host().toString();
+
+    WebLocalFrameLoaderClient_RELEASE_LOG(ResourceMonitoring, "didExceedNetworkUsageThreshold host=%" PUBLIC_LOG_STRING, host.utf8().data());
```
Definitely not OK to print out a `host` with `PUBLIC_LOG_STRING`.

cdumez replied:
We can use `url.host().utf8().data()`

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -2038,6 +2038,10 @@ void WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold()
     if (url.isEmpty())
         return;
 
+    auto host = url.host().toString();
```
Sad to be creating a new String for no good reason now.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
```diff
@@ -2051,7 +2055,7 @@ void WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold()
         action(true);
     else
-        WebProcess::singleton().ensureNetworkProcessConnection().protectedConnection()->sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::ShouldOffloadIFrameForHost(url.host().toStringWithoutCopying()), WTFMove(action), 0);
+        WebProcess::singleton().ensureNetworkProcessConnection().protectedConnection()->sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::ShouldOffloadIFrameForHost(host), WTFMove(action), 0);
```
We can keep the code as is. It was more efficient.

basuke committed:
[ResourceMonitor] Add more meaningful release log information part 2.
https://bugs.webkit.org/show_bug.cgi?id=288174
rdar://145259890

Reviewed by Chris Dumez.

Adding new log category and move current release logs to that category. And also add more meaningful information
for the record.

This is part 2, targeting WebKit.

* Source/WebCore/loader/ResourceMonitor.cpp:
(WebCore::ResourceMonitor::networkUsageLevel const):
* Source/WebCore/loader/ResourceMonitor.h:
* Source/WebCore/loader/ResourceMonitorChecker.cpp:
(WebCore::ResourceMonitorChecker::checkEligibility):
(WebCore::ResourceMonitorChecker::setContentRuleList):
(WebCore::ResourceMonitorChecker::finishPendingQueries):
(WebCore::ResourceMonitorChecker::setNetworkUsageThreshold):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::shouldOffloadIFrameForHost):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::resourceMonitorThrottler):
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::ResourceMonitorURLsController::prepare):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::platformLoadResourceMonitorRuleList):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::loadOrUpdateResourceMonitorRuleList):
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
(WebKit::WebResourceLoader::updateBytesTransferredOverNetwork):
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::didExceedNetworkUsageThreshold):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setResourceMonitorContentRuleList):
(WebKit::WebProcess::setResourceMonitorContentRuleListAsync):

Canonical link: https://commits.webkit.org/292052@main
